tests/tests/integration_tests.rs

```diff
@@ -373,3 +373,139 @@ async fn ws_batch_works() {
+	assert!(res.status().is_client_error());
+
+	// We use the correct content-type
+	let req = Request::builder()
+		.method(Method::POST)
+		.uri(&uri)
+		.header("content-type", "application/json")
+		.body(Body::from(r#"{ "jsonrpsee": "2.0", method: "say_hello", "id": 1 }"#))
+		.expect("request builder");
+
+	let res = http_client.request(req).await.unwrap();
+	assert!(res.status().is_success());
+}
+
+#[tokio::test]
+async fn http_cors_preflight_works() {
+	use hyper::{Body, Client, Method, Request};
+	use jsonrpsee::http_server::AccessControlBuilder;
+
+	let acl = AccessControlBuilder::new().set_allowed_origins(vec!["https://foo.com"]).unwrap().build();
+	let (server_addr, _handle) = http_server_with_access_control(acl).await;
+
+	let http_client = Client::new();
+	let uri = format!("http://{}", server_addr);
+
+	// First, make a preflight request.
+	// See https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS#preflighted_requests for examples.
+	// See https://fetch.spec.whatwg.org/#http-cors-protocol for the spec.
+	let preflight_req = Request::builder()
+		.method(Method::OPTIONS)
+		.uri(&uri)
+		.header("host", "bar.com") // <- host that request is being sent _to_
+		.header("origin", "https://foo.com") // <- where request is being sent _from_
+		.header("access-control-request-method", "POST")
+		.header("access-control-request-headers", "content-type")
+		.body(Body::empty())
+		.expect("preflight request builder");
+
+	let has = |v: &[String], s| v.iter().any(|v| v == s);
+
+	let preflight_res = http_client.request(preflight_req).await.unwrap();
+	let preflight_headers = preflight_res.headers();
+
+	let allow_origins = comma_separated_header_values(&preflight_headers, "access-control-allow-origin");
+	let allow_methods = comma_separated_header_values(&preflight_headers, "access-control-allow-methods");
+	let allow_headers = comma_separated_header_values(&preflight_headers, "access-control-allow-headers");
+
+	// We expect the preflight response to tell us that our origin, methods and headers are all OK to use.
+	// If they aren't, the browser will not make the actual request. Note that if these `access-control-*`
+	// headers aren't return, the default is that the origin/method/headers are not allowed, I think.
+	assert!(preflight_res.status().is_success());
+	assert!(has(&allow_origins, "https://foo.com") || has(&allow_origins, "*"));
+	assert!(has(&allow_methods, "post") || has(&allow_methods, "*"));
+	assert!(has(&allow_headers, "content-type") || has(&allow_headers, "*"));
+
+	// Assuming that that was successful, we now make the actual request. No CORS headers are needed here
+	// as the browser checked their validity in the preflight request.
+	let req = Request::builder()
+		.method(Method::POST)
+		.uri(&uri)
+		.header("host", "bar.com")
+		.header("origin", "https://foo.com")
+		.header("content-type", "application/json")
+		.body(Body::from(r#"{ "jsonrpsee": "2.0", method: "say_hello", "id": 1 }"#))
```

Comment:
? :)

```suggestion
		.body(Body::from(r#"{ "jsonrpc": "2.0", method: "say_hello", "id": 1 }"#))
```

Reply:
Whoops! I guess the server didn't actually care :D